### PR TITLE
Allow Salsa to run with different Visual Studio versions

### PR DESCRIPTION
**Pull request by forki**
I don't have VS 14.0 installed a had trouble to run the VS intergration tests. With this change we try to find a VS installation and use the latest that we find. It's not perfect, but it's good enough to run some basic tests.

### CODE REVIEW

**Comment by dsyme** [comment]
(Yes please! I also normally do compiler work on a laptop without VS 14.0)

**Comment by latkin** [comment]
The VS 14.0 preview is free to install, what is preventing you from using it?

F# 4.0 is shipping in 14.0.  If you want to do F# 4.0 IDE development, it does not seem unreasonable to require you to dev and test on the VS version it will ship with (once a public version is available, of course).

**Comment by forki** [comment]
no way! ;-) I will never again install a beta of VS

**Comment by forki** [comment]
That said: I always assumed F# 4.0 should also work in older VS instances?!

**Comment by latkin** [comment]
If you are doing VS development, you need to be willing to work with pre-release VS.

F# 4.0 is targeting VS 14.0, there are no plans to back-release it on older versions.

**Comment by forki** [comment]
Honestly I don't want to do VS development. I just wanted to make some tests pass in order to find an issue further down in the compiler. As I said before I can't install VS beta since experience says this will fuck up my settings and influence my day work. I'm sure I'm not alone ;-)

**Comment by ctaggart** [comment]
@forki It is wise to put them on their own virtual machines. I'm running the VS beta on a Windows 10 VM. :)

File: vsintegration/src/Salsa/VsMocks.fs [hunk 1]
```diff
@@ -1635,10 +1635,25 @@ module internal VsActual =
...
+
+        let v1 = findVSInstallDirInRegistry "14.0"
+        if isGood v1 then v1 else
+        let v2 = findVSInstallDirInRegistry "12.0"
+        if isGood v2 then v2 else
+        let v3 = findVSInstallDirInRegistry "11.0"
+        if isGood v3 then v3 else
+        let v4 = findVSInstallDirInRegistry "10.0"
+        if isGood v4 then v4 else
+        failwith "could not detect Visual Studio installation path in registry."
```
Comment (enricosada) [inline vsintegration/src/Salsa/VsMocks.fs]: can be refactored to common pattern with something like:
```
let isGood path = .. then Some path else None
 
["14.0"; "12.0"; "11.0"; "10.0"]
|> List.tryPick (findVSInstallDirInRegistry >> isGood) 
|> function Some p -> p | None -> failwith "could not"
```

**Comment by KevinRansom** [comment]
This is not an approach we are comfortable with.  The vsintegration tree of the fsharp4 repo is only expected to work with VisualStudio 2014.  If you can make it appear to work then that is a fortunate accident.  In fact it probably does work but there is no guarantee that it will continue to work, and so we will be declining this request.

VS 2014 is a free preview and after it is released there will be a community edition of it still free and it has support for F# 4.0.  It seems to me like there is not much reason not install it.  Unless disk space is cramped I suppose.

Try it out, you'll like it.
Kevin

**Comment by forki** [comment]
It's not about disk space. After release I'll install it.

If you don't merge it, it's fine. Then I'll just maintain it in my fork and use it to get things done. I'll always have to cherry-picked but that's OK.

**Comment by KevinRansom** [comment]
I'm sorry about that my friend. We believe it's important to dogfood our product, if we don't run the software we work on then how do we really know if it's any good.  And to be clear Steffen you are a huge part of F# 4.0, I would hope you would want to dogfood the product you work on.

**Comment by latkin** [comment]
Kevin and I agree here. We are certainly sympathetic to the pains of working with pre-release installs of VS, but the solution is not to just ignore the version that we are ultimately targeting.  Once there is a free public preview available, that should be the primary target for build and test.

You are of course welcome to add whatever workarounds you need in your own environment.

**Pull request closed without merging**